### PR TITLE
Startup related fixes

### PR DESCRIPTION
**Pull request by GeorgeTsagk**
This PR fixes some startup related bugs that were discovered when integrating tapd with litd.

Read more [here](https://github.com/lightninglabs/taproot-assets/issues/1122)

Closes https://github.com/lightninglabs/taproot-assets/issues/1122

### CODE REVIEW

File: server.go [hunk 1]
```diff
@@ -702,11 +702,16 @@ func (s *Server) waitForReady() error {
...
-		return nil
-
 	case <-s.quit:
 		return fmt.Errorf("tapd is shutting down")
+
+	default:
+		select {
+		case <-s.ready:
+			return nil
+		case <-s.quit:
```
Comment (guggero) [inline server.go:712]: I think this doesn't solve the `tapd: don't set initialized=true when startup (StartAsSubserver) fails` part of the issue. We don't close the `quit` channel when we fail to start there (which is the main entrypoint for `litd`).

File: rpcperms/interceptor.go [hunk 2]
```diff
@@ -413,6 +413,12 @@ func (r *InterceptorChain) rpcStateUnaryServerInterceptor() grpc.UnaryServerInte
 	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo,
 		handler grpc.UnaryHandler) (interface{}, error) {
 
+		if info.Server == nil {
+			return nil, fmt.Errorf("server is nil, can't check " +
+				"RPC state")
+
```
Comment (guggero) [inline rpcperms/interceptor.go:419]: nit: linter complains about superfluous empty line.

File: rpcperms/interceptor.go [hunk 3]
```diff
@@ -413,6 +413,12 @@ func (r *InterceptorChain) rpcStateUnaryServerInterceptor() grpc.UnaryServerInte
 	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo,
 		handler grpc.UnaryHandler) (interface{}, error) {
 
+		if info.Server == nil {
+			return nil, fmt.Errorf("server is nil, can't check " +
```
Comment (guggero) [inline rpcperms/interceptor.go:417]: Can probably be a bit less specific in the error message here, perhaps just "cannot handle call, server not ready"?

File: server.go [hunk 4]
```diff
@@ -702,11 +702,16 @@ func (s *Server) waitForReady() error {
 	// this part of the code, the called component will handle the quit
 	// signal.
 	select {
-	case <-s.ready:
-		return nil
-
 	case <-s.quit:
 		return fmt.Errorf("tapd is shutting down")
+
+	default:
```
Comment (guggero) [inline server.go:708]: Nice fix. This probably explains one class of potential errors we had.

File: server.go [hunk 5]
```diff
@@ -702,11 +702,16 @@ func (s *Server) waitForReady() error {
...
+		select {
+		case <-s.ready:
```
Comment (guggero) [inline server.go:710]: Maybe mention here that we now _wait_ until the server becomes ready or shuts down (meaning this _is_ blocking).

File: server.go [hunk 6]
```diff
@@ -824,17 +824,25 @@ func (s *Server) Name() protofsm.EndpointName {
 // CanHandle returns true if the target message can be routed to this endpoint.
 //
 // NOTE: This method is part of the protofsm.MsgEndpoint interface.
-func (s *Server) CanHandle(msg protofsm.PeerMsg) bool {
-	<-s.ready
+func (s *Server) CanHandle(msg protofsm.PeerMsg) (bool, error) {
+	err := s.waitForReady()
```
Comment (guggero) [inline server.go:828]: Depending on the discussion on the `lnd` PR it _might_ be okay to just return `false` if `s.waitForReady()` returns an error. But calling `s.waitForReady()` should definitely be done so we don't block indefinitely.

**Review by guggero — COMMENTED** [review]
Nice fixes, thanks. Needs some additional checks though.

File: server.go (hunk 1 above)
Comment (GeorgeTsagk) [reply server.go:712]: Ah right, someone waiting at the `waitForReady` endpoint will stay hanging

Will enhance the `StartAsSubserver` method to conditionally close the `s.quit` chan, if initialization faced any err

File: server.go (hunk 6 above)
Comment (GeorgeTsagk) [reply server.go:828]: Right, we don't really want unnecessary error decorations in this interface.

Instead can just log a debug msg and return false

**Review by Roasbeef — APPROVED** [review]
LGTM 🎨

**Pull request merged**